src/object/cli_ec.c

```diff
@@ -394,6 +523,159 @@ ec_has_parity_cli(daos_iod_t *iod)
 	return iod->iod_recxs[0].rx_idx & PARITY_INDICATOR;
 }
 
+static int
+ec_set_head_fetch_params(struct ec_fetch_params *head, daos_iod_t *iods,
+			 unsigned int nr, unsigned int cnt)
+{
+	unsigned int i;
+
+	D_ALLOC_ARRAY(head->iods, nr);
+	if (head->iods == NULL)
+		return -DER_NOMEM;
+	for (i = 0; i < cnt; i++) {
+		head->iods[i] = iods[i];
+		head->nr++;
+	}
+	return 0;
+}
+
+static int
+ec_iod_stripe_cnt(daos_iod_t *iod, struct daos_ec_attr ec_attr)
+{
+	unsigned int	len = ec_attr.e_len;
+	unsigned short	k = ec_attr.e_k;
+	unsigned int	ss = len * k;
+	unsigned int	i;
+	unsigned int	total_stripe_cnt = 0;
+
+	if (iod->iod_type == DAOS_IOD_SINGLE)
+		return iod->iod_nr;
+
+	for (i = 0; i < iod->iod_nr; i++) {
+		uint64_t start = iod->iod_recxs[i].rx_idx * iod->iod_size;
+		uint64_t rem = iod->iod_recxs[i].rx_nr * iod->iod_size;
+		uint64_t partial = start % ss ? ss - start % ss : 0;
+		uint32_t stripe_cnt = 0;
+
+		if (partial && partial < rem) {
+			rem -= partial;
+			stripe_cnt++;
+		}
+		stripe_cnt += rem / ss;
+		if (rem % ss) {
+			stripe_cnt++;
+		}
+		total_stripe_cnt += stripe_cnt;
+	}
+	return total_stripe_cnt;
+}
+
+static int
+ec_update_fetch_params(struct ec_fetch_params *params, daos_iod_t *iod,
+		       struct daos_ec_attr ec_attr, int stripe_cnt)
+{
+	unsigned int	 len = ec_attr.e_len;
+	unsigned short	 k = ec_attr.e_k;
+	unsigned int	 ss = len * k;
+	unsigned int	 i;
+	int		 rc = 0;
+
+	D_ALLOC_ARRAY(params->niod.iod_recxs, stripe_cnt);
+	if (params->niod.iod_recxs == NULL)
+		return -DER_NOMEM;
+	for (i = 0; i < iod->iod_nr; i++) {
+		uint64_t rem = iod->iod_recxs[i].rx_nr * iod->iod_size;
+		uint64_t start = iod->iod_recxs[i].rx_idx * iod->iod_size;
+		uint64_t partial = start % ss ? ss - start % ss : 0;
+
+		if (partial && partial < rem) {
+			params->niod.iod_recxs[params->niod.iod_nr].rx_nr
+				= partial/iod->iod_size;
+			params->niod.iod_recxs[params->niod.iod_nr++].rx_idx
+				= start/iod->iod_size;
```

Comment:
much of this looks eerily similar to the code above in iod_stripe_cnt.  Can some refactoring be done to avoid duplication?

Reply:
Stripe count is determined, and then it is used guide the splitting of recx entries only if it is needed.